shared/ortcustomops.cc

```diff
@@ -1,11 +1,32 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include <mutex>
 #include <set>
 
 #include "onnxruntime_extensions.h"
 #include "ocos.h"
 
+struct OrtCustomOpDomainDeleter {
+  explicit OrtCustomOpDomainDeleter(const OrtApi* ort_api) {
+    ort_api_ = ort_api;
+  }
+  void operator()(OrtCustomOpDomain* domain) const {
+    ort_api_->ReleaseCustomOpDomain(domain);
+  }
+
+  const OrtApi* ort_api_;
+};
+
+using OrtCustomOpDomainUniquePtr = std::unique_ptr<OrtCustomOpDomain, OrtCustomOpDomainDeleter>;
+static std::vector<OrtCustomOpDomainUniquePtr> ort_custom_op_domain_container;
+static std::mutex ort_custom_op_domain_mutex;
+
```

Comment:
Please think it again, why do you need this and when the destructors of the variables will be called?

Reply:
ort_custom_op_domain_container is defined as static, so the destructor will be called when the process exits

Reply:
Well if you have crash, I guess you don't care about memory leaks ;-) The OS will take care of freeing all the memory allocated by the process.
Regarding the order of destruction of global static variables, the standard says (From § 3.6.3 of the C++03 standard):

>     Destructors (12.4) for initialized objects of static storage duration (declared at block scope or at namespace scope) are called as a result of returning from main and as a result of calling exit (18.3). These objects are destroyed in the reverse order of the completion of their constructor or of the completion of their dynamic initialization. If an object is initialized statically, the object is destroyed in the same order as if the object was dynamically initialized. For an object of array or class type, all subobjects of that object are destroyed before any local object with static storage duration initialized during the construction of the sub- objects is destroyed.

Reply:
Agree, they are static so it shouldn't be accessible to any variables not in this file.

Reply:
@wenbingl, wait for crashes.

Reply:
In my opinion, This PR may create more problems than what is solving.

Reply:
What's the risk in your opinion?
I can't see how this changes can create a problem: it's just a static vector where we keep track of all the created domains. The mutex is protecting a simple push operation that won't create any deadlock or similar. And the destructors will only be called at process exit.

Reply:
OrtCustomOpDomainDeleter takes an OrtApi pointer and hold it here, which means this custom op DLL needs to take dependency on the core onnxruntime.dll. And the core onnxruntime.dll must be alive before the custom op dll is unloaded. Before your change, there wasn't such a requirement. Your change added this. And operating systems do not guarantee such an order.

Reply:
So it may crash at process exit.

Reply:
See my comments at https://github.com/microsoft/onnxruntime/pull/11566#issuecomment-1130767041

Reply:
@snnn , your concern is valid. But for the custom op DLL, it should be unloaded before onnxruntime shutdown if the user wants a graceful shutdown. Not sure the onnxruntime environment will unload the DLLs on its destructor, if not, it should be added as well.

Reply:
@snnn in my opinion this specific case is safe, let me explain:
OrtApi is defined as a pointer to a "static constexpr" variable, see https://github.com/microsoft/onnxruntime/blob/006597b9b8d581c4db5e4294863a33c65cb49fb5/onnxruntime/core/session/onnxruntime_c_api.cc#L2553 and https://github.com/microsoft/onnxruntime/blob/006597b9b8d581c4db5e4294863a33c65cb49fb5/onnxruntime/core/session/onnxruntime_c_api.cc#L2274

According to the standard, static initialization is always performed before any dynamic initialization takes place (see https://stackoverflow.com/questions/24531186/is-order-of-static-initialization-implementation-defined). This implies that OrtApi will always be destroyed after all the dynamically initiazed variables are destroyed and therefore its use is perfectly safe.

Reply:
discussed with @snnn offline, we will merge PR and do modifications if there is any concrete example causing the crash.